### PR TITLE
4.x data migration

### PR DESCRIPTION
**Pull request by dianabarsan**
Adds comprehensive guide about how to use https://github.com/medic/couchdb-migration tool to migrate CouchDb from CHT 3.x to CHT 4.x with minimal downtime - similar to any other upgrade, no risk of users being logged out or needing to resync. 

The migration must be performed by a technical person. The process can be repeated and practiced.

### CODE REVIEW

**Comment by dianabarsan** [comment]
I would be super useful to get some feedback on this. Thanks!

File: content/en/apps/guides/hosting/4.x/upgrading.md
Comment (garethbowen) [inline content/en/apps/guides/hosting/4.x/upgrading.md:14]: I'd remove it. Operationally it's very difficult to organise so if it's not necessary we shouldn't suggest it.

File: content/en/apps/guides/hosting/4.x/upgrading.md
Comment (garethbowen) [inline content/en/apps/guides/hosting/4.x/upgrading.md:47]:
```suggestion
To minimize downtime when upgrading, it's advised to prepare the 3.x installation for the 4.x upgrade, and pre-index all views that are required by 4.x.
```

File: content/en/apps/guides/hosting/4.x/upgrading.md
Comment (garethbowen) [inline content/en/apps/guides/hosting/4.x/upgrading.md:53]:
```suggestion
docker-compose run couch-migration pre-index-views <desired CHT version>
```

File: content/en/apps/guides/hosting/4.x/upgrading.md
Comment (garethbowen) [inline content/en/apps/guides/hosting/4.x/upgrading.md:83]: Add some instructions here to help people choose, but also, explain that they should pick either single node or multi node... at the moment it looks like you should do both!

**Review by garethbowen — CHANGES_REQUESTED** [review]
Very comprehensive! Thanks.

Some suggestions inline.

In general I wouldn't hardcode 4.1.0 anywhere (I've added one suggestion along those lines but there are other places where it's hardcoded too). We want to encourage people to always go to the latest.

**Comment by garethbowen** [comment]
We should make sure this gets ATed before merge too!

File: content/en/apps/guides/hosting/4.x/upgrading.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/upgrading.md]: I added something, but with a TODO to link to a proper guide. I brought this up in our last infra call, but I don't think we ended up creating an issue. Do you think this information should live in this page or elsewhere?>

**Review by garethbowen — APPROVED** [review]
Amazing, thank you!

Please update the [4.0.0 release notes](https://docs.communityhealthtoolkit.org/core/releases/4.0.0/) to reference this doc in this same PR - right now there's a "coming soon" link...

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:116]: the "wait until it is up" part - since we're not passing `-d` to be detached, will this naturally stop on it's own or will they need to start the next step in another terminal?  We should explicitly tell them what they're waiting for.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:126]: same comment as above - we should  qualify how they know `move-node` and `verify` calls correctly executed.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:127]:
```suggestion
If all checks pass, proceed with starting CHT-Core 4.x, using the same environment variables you saved in `~/couchdb-single/.env`.   
```

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:138]: Seems worth an example (though I had trouble formatting code inside a suggestion :disappointed: ) :

```
b) Create a data folder for every one of the CouchDb nodes. If you were going to a 3 cluster node, this would be:

   mkdir -p ~/couchdb-cluster/node1
   mkdir -p ~/couchdb-cluster/node2
   mkdir -p ~/couchdb-cluster/node3

```

(Ah - i see below you call it `couchdb_data_main` vs `node1`.  I like `main` as it's less ambiguous which is the main - which ever you want to standardize on!)

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:139]: This seems ambiguous to me:

> You can tell which is your main node by checking your CouchDb docker compose files

Given the importance of this step, I would switch to something like:

```suggestion
c) Copy the 3.x CouchDb data folder into `~/couchdb-cluster/node1`, which will be your main CouchDb node. This main node will create your cluster and the other secondary nodes will be added to it. In `node1`'s environment variable file, define  `CLUSTER_PEER_IPS` . In all other secondary nodes, declare the `COUCHDB_SYNC_ADMINS_NODE` variable instead.  
```

But I'm not sure if that's technically accurate!

Speaking of which - we should specify where you declare these variables.  It seems like each node has it's own `.env` file, but we've only made one in `~/couchdb-cluster/.env`  below. I don't know where to put `CLUSTER_PEER_IPS` vs where to put `COUCHDB_SYNC_ADMINS_NODE` .

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:234]:
```suggestion
i) Change metadata to match the new shard distribution. We declared `$shard_matrix` in step "g" above, so it is still set now: 
```

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:22]: No, this one doesn't require any env vars, as it doesn't run anything,

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:64]: The tool has a command `get-env` which will get these values, and it's explained right below. 
I switched the phrases around so we start with the description of the command and then the description of the variables. 
I don't want to provide instructions on getting them manually.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:66]: > Turning off nginx and haproxy can break the API which we may need up

API can be down for these operations, we interface directly with CouchDb. Definitely the server has to be stopped. Is that not obvious?

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:84]: I think the information about advantages/disadvantages of clustering should be somewhere else, outside of this guide. Hence the "insert link".

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:96]: I would trust that people will know which version of CHT CouchDb they want to install. Serving them with latest image might not be desirable.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:116]: It naturally stops on its own. All of the commands stop on their own. 
How would you suggest we explicitly tell them what they're waiting for?

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:126]: Suggestions appreciated :) 
The command just completes without error at the moment. Should we have a success message at the end?

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:139]: > It seems like each node has it's own .env file, 

This will depend entirely on how we instruct people to self host clusters, which is now unknown?

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md]: I think I changed all things to <place holder>. Please flag if I missed one.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:81]: Fair enough - we're waiting for this [ticket then](https://github.com/medic/cht-docs/issues/831)? Works for me!

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:93]: I can see not going latest - makes sense! 

I think we should make this be `<desired CHT version>` then so it matches the version specified in step 2 for `pre-index-views`.  Forcing them to choose so they get it right seems better instead of deciding for them and they might get it wrong.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:113]: I suggest:

```suggestion
c) Start 4.x CouchDb and wait until it is up. You'll know it is up when the `docker-compose` call exits without errors:
```shell
cd ~/couchdb-single/ 
docker-compose up -d
cd ~/couchdb-migration/ 
docker-compose run couch-migration check-couchdb-up
```
```

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:123]: OMG - these two and `check-couchdb-up` all having a "$command complete" (or something) at the end would be dreamy!  If it's not too late, that'd be my vote.  Then we could have:

> Verify that the migration was successful burn running this and waiting for it to exit with `check-couchdb-up complete`

Otherwise, like above we did:

> Verify that the migration was successful burn running this `docker-compose` command and waiting for it to exit with no errors:`

**Review by esthermmoturi — APPROVED** [review]
Good job Diana

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:74]: I see this was added - thank you!

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:84]: Noting we should come back and add a link to the "this guide" text - fine to leave as is for now - resolving.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:96]: this still seems like a worthwhile change!

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md:126]: I believe success messages were added in #medic/couchdb-migration/pull/7, so we can now update this section with the corresponding messages

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md]: I see this was added - resolving!

**Review by mrjones-plip — APPROVED** [review]
There's two easy changes left which don't need any final sign off. Approving so we can make these last changes an push this liiiiiiivvveee \o/

Excellent work on this!  And thanks for being receptive to my feedback.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md:96]: Oh, so sorry I missed this!

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [reply content/en/apps/guides/hosting/4.x/data-migration.md]: i see these were added - thanks!

**Review by mrjones-plip — APPROVED** [review]
Thanks again for the most recent round of changes! approving again just to be clear - lgtm!

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:22]: We need to specify the raw URL:

```suggestion
curl -s -o ./docker-compose.yml https://raw.githubusercontent.com/medic/couchdb-migration/main/docker-compose.yml
```

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:78]: Having run through the guide now once (almost!), I think we can be a bit more explicit here:

```suggestion
Some CouchDb configuration values must be ported from existent CouchDb to the 4.x installation. Store them in a safe location before shutting down 3.x CouchDb.
Use the migration tool to obtain these values:
```shell
cd ~/couchdb-migration/ 
docker-compose run couch-migration get-env
```

Four values will be output. You already know the user (`COUCHDB_USER`) and password (`COUCHDB_PASSWORD`), but the two you need to save for later are:

* `COUCHDB_SECRET` -  CouchDB secret - Used in encrypting all CouchDB passwords and session tokens.
* `COUCHDB_UUID` -  `CouchDB server UUID - Used in generating replication checkpointer documents, which track where replication progress between every client and the server, and ensure that clients don't re-download or re-upload documents.
```

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/data-migration.md:44]: Per our chat on Slack today, these all need to be `https`, correct?  I think it's worth updating the URLs to be `https` and explicitly call out that the CouchDB Migration utility needs to talk over `https` when talking to Medic OS.

**Review by lorerod — CHANGES_REQUESTED** [review]
Thanks, @dianabarsan; this is getting beautiful. I just added a few suggestions. Let me know what you think.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (dianabarsan) [reply content/en/apps/guides/hosting/4.x/data-migration.md]: Changed.

File: content/en/apps/guides/hosting/4.x/data-migration.md
Comment (lorerod) [reply content/en/apps/guides/hosting/4.x/data-migration.md:29]: In b) you mean: don't define CHT_NETWORK and set COUCH_URL to https://<local_network_ip>
I think option b is more friendly.

**Review by lorerod — APPROVED** [review]
Thanks!

**Pull request merged**
